samples/nrf9160/lte_ble_gateway/src/main.c

```diff
@@ -146,34 +151,77 @@ void nrf_modem_recoverable_error_handler(uint32_t err)
 	error_handler(ERROR_MODEM_RECOVERABLE, (int)err);
 }
 
-/**@brief Callback for GPS events */
-static void gps_handler(const struct device *dev, struct gps_event *evt)
+/**@brief Request assisted GPS data from nRF Cloud. */
+static void agps_request(struct k_work *work)
+{
+	int retval;
+	struct nrf_modem_gnss_agps_data_frame agps_data;
+
+	if (!nrf_cloud_agps_request_in_progress()) {
+		retval = nrf_modem_gnss_read(&agps_data, sizeof(agps_data),
+					     NRF_MODEM_GNSS_DATA_AGPS_REQ);
+		if (retval) {
+			LOG_ERR("Failed to read AGPS data, err %d", retval);
+			return;
+		}
+
+		retval = nrf_cloud_agps_request(&agps_data);
+		if (retval) {
+			LOG_ERR("Failed to request AGPS data, err %d", retval);
+			return;
+		}
+
+		agps_state = AGPS_STATE_REQUESTED;
+	}
+}
+
+/**@brief Callback for GNSS events */
+static void gnss_event_handler(int event)
 {
+	int retval;
+
 	uint32_t button_state, has_changed;
 	struct sensor_data in_data = {
-		.type = GPS_POSITION,
-		.length = evt->nmea.len,
+		.type = GNSS_POSITION,
 	};
 
-	ARG_UNUSED(dev);
-
-	switch (evt->type) {
-	case GPS_EVT_SEARCH_STARTED:
-		printk("GPS_EVT_SEARCH_STARTED\n");
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
 		return;
-	case GPS_EVT_SEARCH_STOPPED:
-		printk("GPS_EVT_SEARCH_STOPPED\n");
+	case NRF_MODEM_GNSS_EVT_FIX:
+		printk("NRF_MODEM_GNSS_EVT_FIX\n");
+		agps_state = AGPS_STATE_SET;
 		return;
-	case GPS_EVT_SEARCH_TIMEOUT:
-		printk("GPS_EVT_SEARCH_TIMEOUT\n");
+	case NRF_MODEM_GNSS_EVT_NMEA:
```

Comment:
I realize now, that `NRF_MODEM_GNSS_EVT_NMEA_FIX` event might have been handy... You get NMEA events once a second after starting GNSS, even when there's no fix. The implementation now always keeps the latest NMEA string stored, but it might not be for a valid fix. This may not be the intention?

GNSS always sends PVT first, immediately followed by the NMEAs. You probably need to read the PVT when you receive an `NRF_MODEM_GNSS_EVT_PVT` event and store in a static/global variable whether it's a valid fix or not. Then you can check whether the fix is valid or not when you receive an `NRF_MODEM_GNSS_EVT_NMEA` event.